pkg/config/configfile.go

```diff
@@ -0,0 +1,369 @@
+package config
+
+import (
+	"errors"
+	"os"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/kubecost/cost-model/pkg/log"
+	"github.com/kubecost/cost-model/pkg/storage"
+)
+
+// HandlerID is a unique identifier assigned to a provided ConfigChangedHandler. This is used to remove a handler
+// from the ConfigFile when it is no longer needed.
+type HandlerID string
+
+//--------------------------------------------------------------------------
+//  ChangeType
+//--------------------------------------------------------------------------
+
+// ChangeType is used to specifically categorize the change that was made on a ConfigFile
+type ChangeType string
+
+// ChangeType constants contain the different types of updates passed through the ConfigChangedHandler
+const (
+	ChangeTypeCreated  ChangeType = "created"
+	ChangeTypeModified ChangeType = "modified"
+	ChangeTypeDeleted  ChangeType = "deleted"
+)
+
+//--------------------------------------------------------------------------
+//  ConfigChangedHandler
+//--------------------------------------------------------------------------
+
+// ConfigChangedHandler is the func handler used to receive change updates about the
+// config file. Both ChangeTypeCreated and ChangeTypeModified yield a valid []byte, while
+// ChangeTypeDeleted yields a nil []byte.
+type ConfigChangedHandler func(ChangeType, []byte)
+
+//--------------------------------------------------------------------------
+//  ConfigFile
+//--------------------------------------------------------------------------
+
+// DefaultHandlerPriority is used as the priority for any handlers added via AddChangeHandler
+const DefaultHandlerPriority int = 1000
+
+// NoBackingStore error is used when the config file's backing storage is missing
+var NoBackingStore error = errors.New("Backing storage does not exist.")
+
+// ConfigFile is representation of a configuration file that can be written to, read, and watched
+// for updates
+type ConfigFile struct {
+	store      storage.Storage
+	file       string
+	dataLock   *sync.Mutex
```

Comment:
Not to be too dismissive, but I don't understand the point. If the `New__` isn't used, then functionally, it doesn't really matter that the mutex is nil. I'm perfectly fine panicking in situations when there's failure to initialize an object type correctly, then attempting to use that object. 

I don't have strong objections to a `sync.Mutex` or embedding a `sync.Mutex` into the type, but I also don't understand a critique against a pointer which is tied to the explicit construction of an object.

Reply:
I suppose I'm objecting to unnecessary usage of a pointer in general, especially given that Go nicely auto-initializes mutexes when set as just `sync.Mutex`. If its used as a guard against initializing the struct incorrectly, I find that a little concerning. I'm definitely not going to demand that this be changed, but I do think this is non-idiomatic.

I totally understand a desire to make sure structs are initialized correctly, and this does feel like a failing of Go sometimes when calling a constructor is _expected_ but cannot be enforced because the struct type has to be exported.

Reply:
It's 100% an allocation strategy, and what I am implying by requiring the use of a `New` func is that at the time the object is allocated, I will also allocate it's fields (unless `nil` is a valid state for a field). This is fundamental to object oriented code. I do understand wanting to implicitly leverage auto allocation/initialization, and I don't have any problem with either approach. 

I'm not sure how you can conclude that it's failing Go when the standard library (and even the language itself) utilizes constructor logic all over the place. Try accessing a map that hasn't been created with `make()` or a `List` that isn't created with `New()` 
https://pkg.go.dev/container/list#New

There are plenty of use cases for stack allocated structures which auto allocate/initialize it's fields, but I find it counterproductive to debate an explicit allocation of a field in a `New` constructor, or the expectation that the object should "just work" when New() is not used. There's a very obvious responsibility of the developer to understand how to allocate and initialize an object. There are only a few ways to do this in go, and we're not breaking outside of the New idiom whatsoever.

Reply:
> I'm perfectly fine panicking in situations when there's failure to initialize an object type correctly, then attempting to use that object.

> the standard library (and even the language itself) utilizes constructor logic all over the place. Try accessing a map that hasn't been created with make() or a List that isn't created with New()

This is a good point. I'm okay with these mutex pointers now. I'm still a little scarred from previous problems with nil panics happening after some duration of the application running, sorry for digging so deep.